model/Versioned.php

```diff
@@ -729,6 +737,101 @@ public function onAfterSkippedWrite() {
 	}
 
 	/**
+	 * Extend permissions to include additional security for objects that are not published to live.
+	 *
+	 * @param Member $member
+	 * @return bool|null
+	 */
+	public function canView($member = null) {
+		// Invoke default version-gnostic canView
+		if ($this->owner->canViewVersioned($member) === false) {
+			return false;
+		}
+	}
+
+	/**
+	 * Determine if there are any additional restrictions on this object for the given reading version.
+	 *
+	 * Override this in a subclass to customise any additional effect that Versioned applies to canView.
+	 *
+	 * This is expected to be called by canView, and thus is only responsible for denying access if
+	 * the default canView would otherwise ALLOW access. Thus it should not be called in isolation
+	 * as an authoritative permission check.
+	 *
+	 * This has the following extension points:
+	 *  - canViewDraft is invoked if Mode = stage and Stage = stage
+	 *  - canViewArchived is invoked if Mode = archive
+	 *
+	 * @param Member $member
+	 * @return bool False is returned if the current viewing mode denies visibility
+	 */
+	public function canViewVersioned($member = null) {
+		// Bypass when live stage
+		$mode = $this->owner->getSourceQueryParam("Versioned.mode");
+		$stage = $this->owner->getSourceQueryParam("Versioned.stage");
+		if ($mode === 'stage' && $stage === static::get_live_stage()) {
+			return true;
+		}
+
+		// Bypass if site is unsecured
+		if (Session::get('unsecuredDraftSite')) {
+			return true;
+		}
+
+		// If we weren't definitely loaded from live, and we can't view non-live content, we need to
+		// check to make sure this version is the live version and so can be viewed
+		$latestVersion = Versioned::get_versionnumber_by_stage($this->owner->class, 'Live', $this->owner->ID);
+		if ($latestVersion == $this->owner->Version) {
+			// Even if this is loaded from a non-live stage, this is the live version
+			return true;
+		}
+
+		// Extend versioned behaviour
+		$extended = null;
+		if($mode === 'stage') {
+			$extended = $this->owner->extendedCan('canViewDraft', $member);
```

Comment:
There's also lots of situations where are _are_ viewing draft, but this check won't be called (because mode != `stage`). For instance, mode can also be `stage_unique`, `latest_versions`, `all_versions`, etc.